libfoundation/src/foundation-java-private.cpp

```diff
@@ -0,0 +1,1353 @@
+        if (!__NextArgument(*t_args, t_range))
+            return false;
+        MCJavaType t_jtype;
+        t_jtype = MCJavaMapTypeCodeSubstring(*t_args, t_range);
+        
+        if (!MCJavaTypeConforms(t_type, t_jtype))
+            return false;
+    }
+    
+    MCTypeInfoRef t_return_type = MCHandlerTypeInfoGetReturnType(p_signature);
+    switch (p_call_type)
+    {
+        case MCJavaCallTypeConstructor:
+            return MCJavaTypeConforms(t_return_type, kMCJavaTypeObject);
+        case MCJavaCallTypeSetter:
+        case MCJavaCallTypeStaticSetter:
+            return MCJavaTypeConforms(t_return_type, kMCJavaTypeVoid);
+        default:
+        {
+            auto t_return_code = static_cast<MCJavaType>(MCJavaMapTypeCode(p_return));
+            return MCJavaTypeConforms(t_return_type, t_return_code);
+        }
+    }
+}
+
+#ifdef TARGET_SUPPORTS_JAVA
+#include <jni.h>
+
+#ifdef TARGET_SUBPLATFORM_ANDROID
+extern JNIEnv *MCJavaGetThreadEnv();
+extern JNIEnv *MCJavaAttachCurrentThread();
+extern void MCJavaDetachCurrentThread();
+#endif
+
+static JNIEnv *s_env;
+static JavaVM *s_jvm;
+
+#if defined(TARGET_PLATFORM_LINUX) || defined(TARGET_PLATFORM_MACOS_X)
+
+static bool s_weak_link_jvm = false;
+
+extern "C" int initialise_weak_link_jvm_with_path(const char*);
+static bool initialise_weak_link_jvm()
+{
+    if (s_weak_link_jvm)
+        return true;
+    
+    MCAutoStringRef t_path;
+#if defined(TARGET_PLATFORM_LINUX)
+    // On linux we require the path to libjvm.so to be in LD_LIBRARY_PATH
+    // so we can just weak link directly.
+    if (!MCStringFormat(&t_path, "libjvm.so"))
+        return false;
+#else
+    const char *t_javahome = getenv("JAVA_HOME");
+    
+    if (t_javahome == nullptr)
+        return false;
+    
+    if (!MCStringFormat(&t_path, "%s/jre/lib/jli/libjli.dylib", t_javahome))
+        return false;
+#endif
+    
+    MCAutoStringRefAsSysString t_jvm_lib;
+    if (!t_jvm_lib . Lock(*t_path))
+        return false;
+    
+    if (!initialise_weak_link_jvm_with_path(*t_jvm_lib))
+        return false;
+    
+    s_weak_link_jvm = true;
+    return true;
+}
+
+#endif
+
+static void init_jvm_args(JavaVMInitArgs *x_args)
+{
+#if defined(TARGET_PLATFORM_LINUX) || defined(TARGET_PLATFORM_MACOS_X)
+    JNI_GetDefaultJavaVMInitArgs(x_args);
+#endif
+}
+
+static bool create_jvm(JavaVMInitArgs *p_args)
+{
+    int ret = 0;
+#if defined(TARGET_PLATFORM_LINUX) || defined(TARGET_PLATFORM_MACOS_X)
+    ret = JNI_CreateJavaVM(&s_jvm, (void **)&s_env, p_args);
+#endif
+    
+    return ret == 0;
+}
+
+bool initialise_jvm()
+{
+#if defined(TARGET_PLATFORM_MACOS_X) || defined(TARGET_PLATFORM_LINUX)
+    if (!initialise_weak_link_jvm())
+        return false;
+    JavaVMInitArgs vm_args;
+    vm_args.version = JNI_VERSION_1_6;
+    init_jvm_args(&vm_args);
+    
+    JavaVMOption* options = new (nothrow) JavaVMOption[1];
+    options[0].optionString = "-Djava.class.path=/usr/lib/java";
```

Comment:
Not in general, but we can add the ability to specify the user class path at a later date.
http://quality.livecode.com/show_bug.cgi?id=19336